# Hybrid NER + Classification Pipeline for OJT Journal Task Tagging
### Deterministic Dictionary + Contextual Transformer NER

This notebook provides an end-to-end, reproducible workflow for the OJT Task Extraction system:
1. **Environment Setup & GPU Acceleration**: PyTorch 2.6.0+cu124, spaCy 3.8.16, transformers.
2. **Real Annotated Data Ingestion**: Cleaned student journal entries (`data/data.jsonl`).
3. **Seed Terms Dictionary**: Seed lookup (`data/terms.csv`) normalized to `IT_TERM` / `CLERICAL_TERM`.
4. **Leak-Free Real Dataset Partitioning**: Sentence-level connected-component splitting (`train.spacy`, `dev.spacy`, `test.spacy`).
6. **7-Check Data Leakage Audit**: Strict validation across splits, unseen benchmark terms, and synthetic near-duplicates (similarity $\ge 0.70$).
7. **Transformer Fine-Tuning**: Patience-based early stopping on GPU for **TRTR** (Real Only).
8. **Side-by-Side Comparative Evaluation**: Testing on real held-out test data and the 65-term unseen benchmark.
9. **Thesis Synthesis & Methodological Findings**: Quantifying generalization lift and documentation.


---
## Phase 1: Environment Setup & GPU Initialization

Import core modules and initialize GPU acceleration on the NVIDIA RTX 3060 Laptop GPU.


In [1]:
import os
import sys
import json
import pandas as pd
import spacy

# Add project root to sys.path
sys.path.insert(0, os.path.abspath("."))

import scripts
from scripts.annotation import (
    load_terms_dictionary,
    find_term_spans,
    load_jsonl,
    report_dataset_diagnostics,
    prepare_real_data_pipeline,
)
from scripts.training import train_ner_trf
from scripts.pipeline import HybridJournalPipeline
from scripts.eval import (
    evaluate_unseen_generalization,
    evaluate_test_docbin,
    evaluate_real_holdout,
    load_unseen_benchmark,
)
from scripts.check_data_leakage import run_all_leakage_checks

# Initialize GPU acceleration
gpu_ready = scripts.init_gpu()
print(f"System Status: GPU Acceleration Active = {gpu_ready}")


/home/caineirb/Documents/PauPau/spaCy-training/.venv/lib/python3.11/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm
[2026-09-23 12:57:30,117] INFO: GPU accelerated with NVIDIA GeForce RTX 3060 Laptop GPU (spacy.require_gpu=True)


System Status: GPU Acceleration Active = True


---
## Phase 2: Load & Inspect Real Annotated Data

Load the authentic, manually annotated OJT student journal dataset (`data/data.jsonl`) and audit its composition diagnostics (negative ratio target: 25–35%, class balance ratio $\le 1.5:1$).


In [2]:
data_path = "data/data.jsonl"
records = load_jsonl(data_path)
print(f"Loaded {len(records)} records from {data_path}\n")

# Report dataset diagnostics (negative ratio, class balance)
diagnostics = report_dataset_diagnostics(records, dataset_label="Real Annotated Dataset")

# Display sample records with highlighted entity spans
print("\nSample Records:")
for r in records[:5]:
    text = r["text"]
    ents = [(text[e["start"]:e["end"]], e["label"]) for e in r.get("entities", [])]
    print(f"  Text    : {text}")
    print(f"  Entities: {ents if ents else '(Negative — 0 entities)'}\n")


[2026-09-23 12:57:30,127] INFO: --- Real Annotated Dataset Diagnostics ---
[2026-09-23 12:57:30,128] INFO:   Total records : 1044
[2026-09-23 12:57:30,128] INFO:   Positives     : 725
[2026-09-23 12:57:30,129] INFO:   Negatives     : 319 (30.6%)
[2026-09-23 12:57:30,129] INFO:   Entity labels : {'CLERICAL_TERM': 528, 'IT_TERM': 514}
[2026-09-23 12:57:30,129] INFO:   Negative ratio is within target range (25-35%).
[2026-09-23 12:57:30,130] INFO:   Class balance acceptable: IT_TERM=514, CLERICAL_TERM=528 (ratio 1.03:1)


Loaded 1044 records from data/data.jsonl


Sample Records:
  Text    : Encoded and organized vendor evaluation data using Microsoft Excel.
  Entities: [('evaluation data', 'CLERICAL_TERM'), ('Microsoft Excel', 'CLERICAL_TERM')]

  Text    : Organized and digitized vendor-related documents.
  Entities: (Negative — 0 entities)

  Text    : Created vendor profiles in the system/database.
  Entities: [('profiles', 'CLERICAL_TERM')]

  Text    : Managed and updated records in the office database.
  Entities: (Negative — 0 entities)

  Text    : Designed identification (ID) cards for vendors/staff.
  Entities: [('cards', 'CLERICAL_TERM')]



---
## Phase 3: Seed Terms Dictionary & Weak Supervision

Inspect `data/terms.csv` containing normalized entity seeds (`IT_TERM` and `CLERICAL_TERM`).


In [3]:
terms_df = pd.read_csv("data/terms.csv")
terms_dict = load_terms_dictionary("data/terms.csv")

print(f"Terms Dictionary: {len(terms_dict)} unique terms loaded.")
print(f"Label distribution:")
print(terms_df["label"].value_counts())

# Test candidate string-matching
# sample_text = "I configured Microsoft Excel for data entry and resolved network issues."
# matched_spans = find_term_spans(sample_text, terms_dict)
# print(f"\nMatching Demo:")
# print(f"  Sentence: {sample_text}")
# print(f"  Matched : {[(sample_text[s[0]:s[1]], s[2]) for s in matched_spans]}")


[2026-09-23 12:57:30,151] INFO: Loaded 291 unique terms from data/terms.csv


Terms Dictionary: 291 unique terms loaded.
Label distribution:
label
IT_TERM          167
CLERICAL_TERM    124
Name: count, dtype: int64


---
## Phase 4: Deduplication & Train/Dev/Test Split (Pure Real Data)

Split `data/data.jsonl` into 70% train, 15% dev, and 15% test partitions using sentence-level connected-component grouping to guarantee **zero sentence leakage** across splits.


In [4]:
pipeline_result = prepare_real_data_pipeline(
    data_jsonl_path="data/data.jsonl",
    output_dir="data/training",
    train_ratio=0.70,
    dev_ratio=0.15,
    seed=42
)

print("\nDocBin files written:")
for split_name, path in pipeline_result["spacy_paths"].items():
    exists = os.path.exists(path)
    size = os.path.getsize(path) if exists else 0
    print(f"  {split_name:<5}: {path} ({size} bytes)")


[2026-09-23 12:57:30,161] INFO: Loaded 1044 records from data/data.jsonl
[2026-09-23 12:57:30,162] INFO: --- Full Dataset Diagnostics ---
[2026-09-23 12:57:30,162] INFO:   Total records : 1044
[2026-09-23 12:57:30,163] INFO:   Positives     : 725
[2026-09-23 12:57:30,163] INFO:   Negatives     : 319 (30.6%)
[2026-09-23 12:57:30,164] INFO:   Entity labels : {'CLERICAL_TERM': 528, 'IT_TERM': 514}
[2026-09-23 12:57:30,164] INFO:   Negative ratio is within target range (25-35%).
[2026-09-23 12:57:30,164] INFO:   Class balance acceptable: IT_TERM=514, CLERICAL_TERM=528 (ratio 1.03:1)
[2026-09-23 12:57:30,166] WARNING: Found 7 duplicate sentence texts with conflicting annotations!
[2026-09-23 12:57:30,167] WARNING:   Conflict on: 'Office supplies management maintained equipment inventory'
[2026-09-23 12:57:30,167] WARNING:   Conflict on: 'User authentication system protected data security'
[2026-09-23 12:57:30,168] WARNING:   Conflict on: 'Data entry and filing tasks'
[2026-09-23 12:57:30,16


DocBin files written:
  train: data/training/train.spacy (92052 bytes)
  dev  : data/training/dev.spacy (22778 bytes)
  test : data/training/test.spacy (23337 bytes)


---
## Phase 5: Synthetic Augmentation

> **Note**: The original paraphrase-based synthetic augmentation pipeline (T5 paraphraser + template generator) has been replaced.
> See `docs/archive/paraphrase_augmentation_methodology.md` for the historical methodology.
> The new LLM-direct generation approach uses `scripts/generate_llm_synthetic.py`.


In [ ]:
# Synthetic augmentation is now handled by scripts/generate_llm_synthetic.py
# See api/README.md or docs/ for instructions.
print("Phase 5: LLM-direct synthetic generation — run scripts/generate_llm_synthetic.py separately.")


---
## Phase 6: 7-Check Data Leakage & Benchmark Isolation Audit

Verify strict zero data leakage across all 7 checks before evaluating models:
1. Cross-Split Document Duplication = 0
2. Cross-Split Sentence Duplication = 0
3. Terms CSV Unseen Benchmark Leakage = 0
4. Training Annotation Unseen Benchmark Leakage = 0
5. EntityRuler Matchability on Unseen Benchmark = 0
6. Real Holdout Data Isolation = 0
7. **Synthetic Pool Isolation**: 0 exact or near-duplicates (similarity $\ge 0.70$) between synthetic pool and evaluation partitions; 0 benchmark terms in synthetic pool.


In [6]:
all_passed = run_all_leakage_checks()
print(f"\nOverall Leakage Audit Result: {'ALL CHECKS PASSED (ZERO LEAKAGE)' if all_passed else 'LEAKAGE DETECTED'}")


[2026-09-23 12:57:30,677] INFO: Loaded 85 unseen benchmark samples from data/test/unseen_benchmark.jsonl
[2026-09-23 12:57:30,678] INFO: GPU accelerated with NVIDIA GeForce RTX 3060 Laptop GPU (spacy.require_gpu=True)
[2026-09-23 12:57:30,691] INFO: Loaded 291 unique terms from data/terms.csv
[2026-09-23 12:57:30,692] INFO: Loading transformer model from 'models/ner_trf/model-best'...



###########################################################################
  DATA LEAKAGE & BENCHMARK ISOLATION AUDIT
###########################################################################

[INFO] Loading dataset splits from data/training/*.spacy...
[INFO] Initializing pipeline for EntityRuler matchability audit...


[2026-09-23 12:57:34,076] INFO: Configured EntityRuler with 291 patterns before NER.
[2026-09-23 12:57:34,145] INFO: Loaded 291 unique terms from data/terms.csv



  Check 1: Cross-Split Document Duplication
[PASS] Zero duplicate documents across train/dev/test splits.
       Docs: train=687, dev=147, test=148

  Check 2: Cross-Split Sentence Duplication
[PASS] Zero duplicate sentences across train/dev/test splits.
       Sentences: train=762, dev=168, test=170

  Check 3: Unseen Benchmark Terms in data/terms.csv
[PASS] Zero of 65 unseen benchmark terms appear in data/terms.csv.
       Dictionary size checked: 291 terms.

  Check 4: Unseen Benchmark Terms in Training Annotations
[PASS] Zero of 65 unseen benchmark terms appear in data/data.jsonl.
       Annotated entity pool: 696 unique entity strings.

  Check 5: EntityRuler Matchability on Unseen Benchmark
[PASS] Zero unseen benchmark terms are matchable as exact entities by the EntityRuler.
       [NOTE] Sub-span token overlap detected for 1 terms:
         - 'Tailwind CSS' contains dictionary token 'CSS' (IT_TERM)

  Check 6: Real Holdout Scaffolding & Data Isolation
[PASS] Real holdout datas

---
## Phase 7: Transformer Fine-Tuning (TRTR)

Fine-tune the RoBERTa-base transformer NER model on the real training partition with patience-based early stopping:
- `max_steps = 2500`
- `eval_frequency = 50`
- `patience = 400`
- `use_gpu = 0`


In [ ]:
trtr_best = "models/ner_trf_trtr/model-best"

# Condition TRTR (Train Real, Test Real)
if not os.path.exists(trtr_best):
    print("Fine-tuning TRTR model on pure real data (data/training/train.spacy)...")
    train_ner_trf(
        output_dir="models/ner_trf_trtr",
        train_path="data/training/train.spacy",
        dev_path="data/training/dev.spacy",
        config_path="config_trf.cfg",
        use_gpu=0,
        max_steps=2500,
        eval_frequency=50,
        patience=400
    )
else:
    print(f"TRTR pre-trained model checkpoint ready: {trtr_best}")


---
## Phase 8: Evaluation (TRTR Baseline)

Evaluate the TRTR model on the real evaluation partitions:
1. **Held-Out Real Test Set** (`data/training/test.spacy` — 148 documents)
2. **Unseen-Term Benchmark** (`data/test/unseen_benchmark.jsonl` — 65 unseen enterprise technologies)
3. **Real-World Holdout** (`data/test/holdout.jsonl` if populated)


In [ ]:
# Initialize pipeline
pipe_trtr = HybridJournalPipeline(model_path=trtr_best, terms_csv_path="data/terms.csv")

# 1. Evaluate on Real Held-Out Test Set
test_trtr = evaluate_test_docbin(pipe_trtr, "data/training/test.spacy")

# 2. Evaluate on Unseen Benchmark (all three modes)
unseen_trtr = evaluate_unseen_generalization(pipe_trtr)

# Print TRTR Results
print("                       TRTR (Real Only) BASELINE RESULTS")
print("=" * 75)
print(f"{'METRIC':<42} | {'TRTR (Real Only)':>18}")
print("-" * 75)
print(f"{'Held-Out Real Test Overall F1':<42} | {test_trtr['overall_f1']:>16}%")
print(f"{'Held-Out Real Test Overall Precision':<42} | {test_trtr['overall_precision']:>16}%")
print(f"{'Held-Out Real Test Overall Recall':<42} | {test_trtr['overall_recall']:>16}%")
print("-" * 75)
for lbl in ['IT_TERM', 'CLERICAL_TERM']:
    sl = test_trtr['labels'].get(lbl, {})
    print(f"  {lbl + ' F1':<40} | {sl.get('f1', 0):>16}%")
    print(f"  {lbl + ' Precision':<40} | {sl.get('precision', 0):>16}%")
    print(f"  {lbl + ' Recall':<40} | {sl.get('recall', 0):>16}%")
print("-" * 75)

b_trf = unseen_trtr['modes']['transformer_only']
print(f"{'Unseen Benchmark Transformer Recall':<42} | {b_trf.get('recall', 0):>16}%")
print(f"{'Unseen Benchmark Transformer Precision':<42} | {b_trf.get('precision', 0):>16}%")
print(f"{'Unseen Benchmark Transformer F1':<42} | {b_trf.get('f1', 0):>16}%")
b_hyb = unseen_trtr['modes']['hybrid']
print(f"{'Unseen Benchmark Hybrid Recall':<42} | {b_hyb.get('recall', 0):>16}%")
print(f"{'Unseen Benchmark Hybrid F1':<42} | {b_hyb.get('f1', 0):>16}%")
print(f"{'Unseen Benchmark (Generalization Lift)':<42} | {unseen_trtr['generalization_lift']:>17}")
print("=" * 75)


---
## Phase 9: Findings & Thesis Methodological Summary

### Baseline Conclusions (TRTR):
The TRTR (Train Real, Test Real) condition establishes the baseline performance using only authentic student journal annotations.

> **Note**: The comparative evaluation against LLM-direct synthetic augmentation (TRSTR-LLM) will be documented separately after generation and retraining.
> For historical reference on the prior paraphrase-based augmentation approach, see `docs/archive/paraphrase_augmentation_methodology.md`.

> [!WARNING]
> **Limitations Disclaimer**: Any synthetic augmentation is introduced to compensate for the limited volume of the authentic annotated corpus. TRTR remains the reference point for real-data-only performance. Augmented results should not be misconstrued as evidence that the system was trained exclusively on authentic data.
